components/script/dom/readablestreamdefaultreader.rs

```diff
@@ -169,45 +165,81 @@ impl ReadableStreamDefaultReader {
             .pop_front()
             .expect("Reader must have read request when remove is called into.")
     }
+
+    fn get_read_requests(&self) -> VecDeque<ReadRequest> {
+        mem::take(&mut *self.read_requests.borrow_mut())
+    }
 }
 
 impl ReadableStreamDefaultReaderMethods for ReadableStreamDefaultReader {
     /// <https://streams.spec.whatwg.org/#default-reader-read>
     fn Read(&self) -> Rc<Promise> {
         let promise = Promise::new(&self.reflector_.global());
 
-        self.stream
-            .perform_pull_steps(ReadRequest::Read(promise.clone()));
+        if let Some(stream) = self.stream.get() {
+            stream.perform_pull_steps(ReadRequest::Read(promise.clone()));
+        } else {
+            promise.reject_error(Error::Type("stream is undefined".to_owned()));
+        }
 
         promise
     }
 
     /// <https://streams.spec.whatwg.org/#default-reader-release-lock>
+    #[allow(unsafe_code)]
     fn ReleaseLock(&self) {
-        if self.stream.is_readable() {
-            self.closed_promise.reject_native(&());
-        }
-
-        // TODO: https://streams.spec.whatwg.org/#readable-stream-reader-generic-release
-
-        // <https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests>
-        for request in self.read_requests.borrow_mut().drain(0..) {
-            // TODO: use TypeError.
-            let cx = GlobalScope::get_cx();
-            rooted!(in(*cx) let mut rval = UndefinedValue());
-            request.error_steps(rval.handle());
+        // step 2
+        // https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreaderrelease
+        if let Some(stream) = self.stream.get() {
+            // step 2.1
+            // https://streams.spec.whatwg.org/#readable-stream-reader-generic-release
+            stream.release_lock(&self.closed_promise);
+            self.stream.set(None);
+
+            // step 2.2
+            let error =
+                Error::Type("No chunks are available because the stream is errored".to_owned());
+
+            // step 2.3
+            // <https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests>
+
+            // step 2.3.1
+            let mut read_requests = self.get_read_requests();
+            // step 2.3.2 & 2.3.3
+            for request in read_requests.drain(0..) {
+                let cx = GlobalScope::get_cx();
+                rooted!(in(*cx) let mut rval = UndefinedValue());
+                unsafe {
+                    error
+                        .clone()
+                        .to_jsval(*cx, &*self.global(), rval.handle_mut())
+                };
+                request.error_steps(rval.handle());
+            }
+        } else {
+            // step 1
```

Comment:
I think we should also follow the spec more closely:

- add a conditional on `self.stream.get().is_none`().
- Call into a new method for https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreaderrelease
- Move rest of algo to that new method(it can be called from other places later).